pkg/runtime/unstructured/converter.go

```diff
@@ -0,0 +1,444 @@
+	st, dt := sv.Type(), dv.Type()
+
+	// Check if the object has a custom JSON marshaller/unmarshaller.
+	if st.Implements(marshalerType) {
+		var data []byte
+		var err error
+		if sv.Kind() == reflect.Ptr && sv.IsNil() {
+			data = []byte("null")
+		} else {
+			marshaler := sv.Interface().(encodingjson.Marshaler)
+			data, err = marshaler.MarshalJSON()
+			if err != nil {
+				return err
+			}
+		}
+		if string(data) == "null" {
+			// We're done - we don't need to store anything.
+		} else {
+			switch {
+			case len(data) > 0 && data[0] == '"':
+				var result string
+				err := json.Unmarshal(data, &result)
+				if err != nil {
+					return fmt.Errorf("error decoding from json: %v", err)
+				}
+				dv.Set(reflect.ValueOf(result))
+			case len(data) > 0 && data[0] == '{':
+				result := make(map[string]interface{})
+				err := json.Unmarshal(data, &result)
+				if err != nil {
+					return fmt.Errorf("error decoding from json: %v", err)
+				}
+				dv.Set(reflect.ValueOf(result))
+			default:
+				var result int64
+				err := json.Unmarshal(data, &result)
+				if err != nil {
+					return fmt.Errorf("error decoding from json: %v", err)
+				}
+				dv.Set(reflect.ValueOf(result))
+			}
+		}
+		return nil
+	}
+
+	switch st.Kind() {
+	case reflect.String, reflect.Bool,
+		reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		if dt.Kind() == reflect.Interface && dv.NumMethod() == 0 {
+			dv.Set(reflect.New(st))
+		}
+		dv.Set(sv)
+		return nil
+	case reflect.Map:
+		return mapToUnstructured(sv, dv)
+	case reflect.Slice:
+		return sliceToUnstructured(sv, dv)
+	case reflect.Ptr:
+		return pointerToUnstructured(sv, dv)
+	case reflect.Struct:
+		return structToUnstructured(sv, dv)
+	case reflect.Interface:
+		return interfaceToUnstructured(sv, dv)
+	default:
+		return fmt.Errorf("unrecognized type: %v", st.Kind())
+	}
+}
+
+func mapToUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+	if dt.Kind() == reflect.Interface && dv.NumMethod() == 0 {
+		dv.Set(reflect.MakeMap(mapStringInterfaceType))
+		dv = dv.Elem()
+		dt = dv.Type()
+	}
+	if dt.Kind() != reflect.Map {
+		return fmt.Errorf("cannot convert struct to: %v", dt.Kind())
+	}
+
+	if !st.Key().AssignableTo(dt.Key()) && !st.Key().ConvertibleTo(dt.Key()) {
+		return fmt.Errorf("cannot copy map with non-assignable keys: %v %v", st.Key(), dt.Key())
+	}
+
+	for _, key := range sv.MapKeys() {
+		value := reflect.New(dt.Elem()).Elem()
+		if err := toUnstructured(sv.MapIndex(key), value); err != nil {
+			return err
+		}
+		if st.Key().AssignableTo(dt.Key()) {
+			dv.SetMapIndex(key, value)
+		} else {
+			dv.SetMapIndex(key.Convert(dt.Key()), value)
+		}
+	}
+	return nil
+}
+
+func sliceToUnstructured(sv, dv reflect.Value) error {
+	dt := dv.Type()
+	if dt.Kind() == reflect.Interface && dv.NumMethod() == 0 {
+		dv.Set(reflect.MakeSlice(reflect.SliceOf(dt), sv.Len(), sv.Cap()))
```

Comment:
Avoid allocating here if the slice is nil.  If the slice is empty... I think we can keep a cache of empty slices and use them (although we'd need to double check a go slice allocation from an empty slice creates a new backing store).